linkerd/opaque-transport/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+use bytes::{
+    buf::{Buf, BufMut},
+    Bytes, BytesMut,
+};
+use linkerd2_dns_name::Name;
+use linkerd2_error::Error;
+use linkerd2_io::{self as io, AsyncReadExt, AsyncWriteExt};
+use linkerd2_proxy_transport::Detect;
+use prost::Message;
+use std::str::FromStr;
+use tracing::trace;
+
+mod proto {
+    include!(concat!(env!("OUT_DIR"), "/opaque.proxy.l5d.io.rs"));
+}
+
+#[derive(Clone, Debug)]
+pub struct Header {
+    /// The target port.
+    pub port: u16,
+
+    /// The logical name of the target (service), if one is known.
+    pub name: Option<Name>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct DetectHeader(());
+
+const PREFACE: &[u8] = b"proxy.l5d.io/opaque\r\n\r\n";
+const PREFACE_AND_SIZE_LEN: usize = PREFACE.len() + 4;
+
+#[async_trait::async_trait]
+impl Detect for DetectHeader {
+    type Protocol = Header;
+
+    #[inline]
+    async fn detect<I: io::AsyncRead + Send + Unpin + 'static>(
+        &self,
+        io: &mut I,
+        buf: &mut BytesMut,
+    ) -> Result<Option<Header>, Error> {
+        Header::read_prefaced(io, buf).await.map_err(Into::into)
+    }
+}
+
+impl Header {
+    pub async fn write(&self, io: &mut (impl io::AsyncWrite + Unpin)) -> Result<usize, Error> {
+        let mut buf = self.encode_prefaced_buf()?;
+        let mut sz = 0usize;
+
+        while !buf.is_empty() {
+            sz += io.write_buf(&mut buf).await?;
+            trace!(written = sz, remaining = buf.len(), "Wrote header");
+        }
+
+        Ok(sz)
+    }
+
+    #[inline]
+    pub fn encode_prefaced_buf(&self) -> Result<Bytes, Error> {
+        let mut buf = BytesMut::new();
+        self.encode_prefaced(&mut buf)?;
+        Ok(buf.freeze())
+    }
+
+    /// Encodes the connection header to a byte buffer.
+    pub fn encode_prefaced(&self, buf: &mut BytesMut) -> Result<(), Error> {
+        buf.reserve(PREFACE_AND_SIZE_LEN);
+        buf.put(PREFACE);
+
+        debug_assert!(buf.capacity() >= 4);
+        // Safety: These bytes must be initialized below once the message has
+        // been encoded.
+        unsafe {
+            buf.advance_mut(4);
+        }
+
+        self.encode(buf)?;
+
+        // Once the message length is known, we back-fill the length at the
+        // start of the buffer.
+        let len = buf.len() - PREFACE_AND_SIZE_LEN;
+        assert!(len <= std::u32::MAX as usize);
+        {
+            let mut buf = &mut buf[PREFACE.len()..PREFACE_AND_SIZE_LEN];
+            buf.put_u32(len as u32);
+        }
```

Comment:
hmm, actually...isn't all this logic just `prost`'s [`Message::encode_length_delimited`](https://docs.rs/prost/0.6.1/prost/trait.Message.html#method.encode_length_delimited) method? is there a reason we're not just using that?

Reply:
The length encoded by this is itself [variable-length](https://docs.rs/prost/0.6.1/src/prost/message.rs.html#70); and so it complicates the read-side since we don't know how much data we need to buffer to decode the length. We want our length field to be a fixed size.

Reply:
Ah, yup, I see. Might be worth a comment explaining that.